docs/how-it-works/internals.md

```diff
@@ -0,0 +1,120 @@
+
+#### Authentication
+
+Since all data is loaded via Python over the Jupyter-provided WebSocket, Lonboard doesn't need to implement any sort of authentication itself. The user is able to use whatever credentials they're familiar with, say for S3, to access data. Those credentials never need to be passed to the browser.
+
+In contrast, having the JavaScript side load data directly from a URL requires some way for users to specify authorization headers or API keys for access to private datasets.
+
+### Layer Creation
+
+With the exception of the `BitmapLayer`, no _direct_ deck.gl layers are used. deck.gl layers are used _exclusively_ through the [`@geoarrow/deck.gl-layers`](https://github.com/geoarrow/deck.gl-layers) glue library. This library connects GeoArrow data to the low-level binary data API of each deck.gl layer. **All** accessors are passed in directly as binary buffers.
+
+This means that in JavaScript the data _never_ leaves an Arrow binary representation. It's parsed from Parquet to Arrow in WebAssembly, the Arrow buffers are copied out of WebAssembly memory, but never parsed to JSON strings or JavaScript objects.
+
+### Accessors and data-driven rendering
+
+Loading data from Python also helps to simplify data-driven rendering.
+
+#### No function serialization
+
+If we can make the data movement between Python and JavaScript fast enough, we can serialize _computed rendering properties_ rather than the function to generate them. Lonboard serializes data, not functions.
+
+Any accessor used to style data is computed in Python, stored as a binary array, serialized to Parquet and sent to the frontend. Other than parsing the input Parquet buffers to Arrow, no data transformations or modifications are done on the JavaScript side.
+
+In Pydeck users had to pass something like
+
+```py
+getRadius="@@=properties.valuePerSqm"
+```
+
+which would be expanded into
+
+```js
+getRadius: (object) => object.properties.valuePerSqm
+```
+
+on the JavaScript side.
+
+I find this syntax to be horrible UX. The user is no longer writing Python, they need to learn a new DSL just for this.
+
+RGB Colors for 3 million points take up 9MiB in memory (or 12MiB for RGBA). In the example of speed test data, 12MiB of RGBA data compressed down to 4MiB for transit. Float32 radii values for 3 million points take up 12MiB in memory and 9MiB compressed.
+
+Even for users on remote Python environments, this takes no more than a couple seconds to download, and then can render almost instantly. For non-point datasets with a fewer number of geometries, the byte size of accessors are even smaller relative to the geometry buffers.
```

Comment:
> a couple seconds to download, and then can render almost instantly

For a user, this is just "how long it takes my viz to update", they don't care if the render is instant if they have to wait seconds for the data. Contrast to applying a color map in JavaScript, which would feel instant (just playing devil's advocate here, I know many users will want to apply their color maps in python, even if it is slow)

Reply:
For local Python environments, applying the colormap on the Python side is just as fast as doing it in JS. For remote environments, I believe the "understandability" of Python code is preferred to the user somehow serializing a colormap to JS and applying it there. I think the benefit of having Lonboard be "just Python" is worth waiting 1-2 seconds for a new colormap when you have 3 million rows.